webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,234 @@
+import { Button, Col, InputNumber, Modal, Row } from 'antd';
+import { Form, Input, Select } from 'antd';
+import { ModalProps } from 'antd/es/modal/Modal';
+import yaml from 'js-yaml';
+import React, { useCallback, useEffect, useState } from 'react';
+import MonacoEditor from 'react-monaco-editor';
+
+import useStorage from 'hooks/useStorage';
+import { getResourcePools, getTemplates } from 'services/api';
+import { NotebookConfig, RawJson, ResourcePool, Template } from 'types';
+import { launchNotebook, previewNotebook } from 'utils/task';
+
+import Link from './Link';
+import RadioGroup from './RadioGroup';
+
+const { Option } = Select;
+const { Item } = Form;
+
+const STORAGE_PATH = 'notebook-launch';
+const STORAGE_KEY = 'notebook-config';
+
+interface Props extends ModalProps {
+  onLaunch?: () => void;
+  visible?: boolean;
+}
+
+const NotebookModal: React.FC<Props> = (
+  { visible = false, onLaunch = (() => null), ...props }: Props,
+) => {
+  const storage = useStorage(STORAGE_PATH);
+  const [ showFullConfig, setShowFullConfig ] = useState(false);
+  const [ templates, setTemplates ] = useState<Template[]>([]);
+  const [ resourcePools, setResourcePools ] = useState<ResourcePool[]>([]);
+  const [ resourceTypeOptions, setResourceTypeOptions ] =
+    useState<{id:string, label:string}[]>([ { id:'CPU', label:'CPU' }, { id:'GPU', label:'GPU' } ]);
+  const [ resourceType, setResourceType ] =
+    useState(storage.getWithDefault(STORAGE_KEY, { type: undefined }).type);
+  const [ form ] = Form.useForm();
+
+  useEffect(() => {
+    const fetchTemplates = async () => {
+      setTemplates(await getTemplates({}));
+    };
```

Comment:
**chore:** let's redefine these fetch functions with `useCallback` hooks outside of the `useEffect` and add try/catch blocks. Otherwise if the api fails, the thrown errors will break the react app.